### PR TITLE
Network LInk Control

### PR DESCRIPTION
**Pull request by apurvabanka**
## Summary by Sourcery

Add support for NetworkLinkControl in the KML library, including its registration, documentation, and testing.

New Features:
- Introduce the NetworkLinkControl class to handle KML network link control elements.

Enhancements:
- Update the kml_children union and RegistryItem to include NetworkLinkControl.

Documentation:
- Add a new KML document for network link control in the documentation.

Tests:
- Add tests for the NetworkLinkControl class to verify its functionality and integration.

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Introduced the `NetworkLinkControl` class to manage network link control features in KML files.
- Extended the `kml_children` union type to include `NetworkLinkControl`.
- Registered `NetworkLinkControl` attributes in the registry for KML handling.
- Added test cases to verify the functionality of the `NetworkLinkControl` class.
- Provided an example KML file demonstrating the use of `NetworkLinkControl`.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>__init__.py</strong><dd><code>Add NetworkLinkControl to module exports</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

fastkml/__init__.py

<li>Added import for <code>NetworkLinkControl</code>.<br> <li> Included <code>NetworkLinkControl</code> in the <code>__all__</code> list.<br>

</details>

  </td>
  <td><a href="https://github.com/cleder/fastkml/pull/392/files#diff-3629e121239e3f7c6ac52604b011106bd7d417fc59a153d5b3d677bbaf7b7653">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>kml.py</strong><dd><code>Integrate NetworkLinkControl into KML handling</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

fastkml/kml.py

<li>Imported <code>NetworkLinkControl</code>.<br> <li> Extended <code>kml_children</code> union type to include <code>NetworkLinkControl</code>.<br> <li> Registered <code>NetworkLinkControl</code> in the KML registry.<br>

</details>

  </td>
  <td><a href="https://github.com/cleder/fastkml/pull/392/files#diff-613ee78b8f0776d39e81909e8b514a8053b408d2dc58eb0830deb146a8e12141">+22/-3</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>network_link_control.py</strong><dd><code>Implement NetworkLinkControl class for KML</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

fastkml/network_link_control.py

<li>Introduced <code>NetworkLinkControl</code> class with several attributes.<br> <li> Registered <code>NetworkLinkControl</code> attributes in the registry.<br>

</details>

  </td>
  <td><a href="https://github.com/cleder/fastkml/pull/392/files#diff-a1f0144f852873c43333cf9c251486b149024f35c4c59ba6d034a4ab8035b43e">+218/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>network_link_control_test.py</strong><dd><code>Add tests for NetworkLinkControl functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/network_link_control_test.py

<li>Added test cases for <code>NetworkLinkControl</code> class.<br> <li> Tested object creation and KML parsing.<br>

</details>

  </td>
  <td><a href="https://github.com/cleder/fastkml/pull/392/files#diff-24d06fbe476ee02851957e7f58bbad7b2a9c5b29d47d53877b2464251e71c6df">+83/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>network.kml</strong><dd><code>Provide example KML for NetworkLinkControl</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docs/network.kml

- Added example KML file demonstrating `NetworkLinkControl`.

</details>

  </td>
  <td><a href="https://github.com/cleder/fastkml/pull/392/files#diff-e79a111f68b0377b90df069563d1e6a04e5bdfd0c7b41c5b01127446c0dfc996">+16/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: fastkml/containers.py [hunk 1]
```diff
@@ -334,6 +410,17 @@ def get_style_by_url(self, style_url: str) -> Optional[Union[Style, StyleMap]]:
         set_element=xml_subelement_list,
     ),
 )
+registry.register(
```
Comment (cleder) [inline fastkml/containers.py:413]: `Schema` is not valid for `NetworkLinkControl`

File: fastkml/features.py [hunk 2]
```diff
@@ -222,6 +225,15 @@ class _Feature(TimeMixin, _BaseObject):
...
     extended_data: Optional[ExtendedData]
+    min_refresh_period: Optional[float] = None,
+    max_session_length: Optional[float] = None,
+    cookie: Optional[str] = None,
+    message: Optional[str] = None,
+    link_name: Optional[str] = None,
+    link_description: Optional[str] = None,
+    link_snippet: Optional[str] = None,
+    expires: Optional[KmlDateTime] = None,
+    update:Optional[Update] = None,
```
Comment (cleder) [inline fastkml/features.py:236]: All these are not valid for a feature

File: fastkml/features.py [hunk 3]
```diff
@@ -244,6 +256,15 @@ def __init__(
...
         extended_data: Optional[ExtendedData] = None,
+        min_refresh_period: Optional[float] = None,
+        max_session_length: Optional[float] = None,
+        cookie: Optional[str] = None,
+        message: Optional[str] = None,
+        link_name: Optional[str] = None,
+        link_description: Optional[str] = None,
+        link_snippet: Optional[str] = None,
+        expires: Optional[KmlDateTime] = None,
+        update:Optional[Update] = None,
```
Comment (cleder) [inline fastkml/features.py:267]: Not valid for feature

File: fastkml/features.py [hunk 4]
```diff
@@ -407,6 +437,95 @@ def __init__(
...
+    _Feature,
+    item=RegistryItem(
+        ns_ids=("kml", ),
+        classes=(KmlDateTime,),
+        attr_name="expires",
+        node_name="expires",
+        get_kwarg=datetime_subelement_kwarg,
+        set_element=datetime_subelement,
+    ),
+)
```
Comment (cleder) [inline fastkml/features.py:527]: see above, not valid for feature

File: fastkml/update.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+class Delete(_XMLObject):
+    def __init__():
+        pass
+
+class Update:
+
+    def __init__(
+            target_href: Optional[str] = None,
+            change: Optional[Change] = None,
```
Comment (cleder) [inline fastkml/update.py:32]: `<Update>` can contain any number of `<Change>`, `<Create>`, and `<Delete>` elements, so this should be an iterable

**Review by cleder — CHANGES_REQUESTED** [review]

File: fastkml/containers.py [hunk 6]
```diff
@@ -323,6 +326,79 @@ def get_style_by_url(self, style_url: str) -> Optional[Union[Style, StyleMap]]:
         )
 
 
+class NetworkLinkControl(_Container):
```
[the remark this answers is not in the document]
Comment (cleder) [reply fastkml/containers.py:329]: I think it should live in its own module to avoid ambiguity

[the remark this answers is not in the document]
Comment (apurvabanka) [reply fastkml/containers.py:329]: That sounds right. Will create a separate class for NetworkLinkControl to inherit from.

File: fastkml/times.py [hunk 7]
```diff
@@ -189,6 +189,11 @@ def __str__(self) -> str:
                 else self.dt.isoformat()
             )
         return self.dt.isoformat()
+    
+    @classmethod
+    def get_tag_name(cls) -> str:
```
Comment (cleder) [inline fastkml/times.py:194]: What is this used for?

File: fastkml/network_control.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+from fastkml.registry import RegistryItem
+from fastkml.base import _XMLObject
+from fastkml.times import KmlDateTime
+from fastkml.update import Update
+from fastkml.views import Camera, LookAt
+
+
+class _NetworkControl(_XMLObject):
+    """
+    Absrtact base class representing NetworkLinkControl in KML.
```
Comment (cleder) [inline fastkml/network_control.py:37]: Why do you need an abstract base class here? NetworkLinkControl is the one and only concrete Implementation.

**Review by cleder — CHANGES_REQUESTED** [review]

File: docs/working_with_kml.rst [hunk 9]
```diff
@@ -232,7 +236,7 @@ Now we can remove the CascadingStyle from the document and have a look at the re
           <kml:displayMode>hide</kml:displayMode>
         </kml:BalloonStyle>
       </kml:Style>
-      <kml:Placemark id="04AFE6060F147CE66FBD">
+      <kml:Placemark id="04SAFE6060F147CE66FBD">
```
Comment (cleder) [inline docs/working_with_kml.rst:239]: this is breaking the doctests please revert

File: fastkml/__init__.py [hunk 10]
```diff
@@ -31,6 +31,7 @@
 from fastkml.atom import Link as AtomLink
 from fastkml.containers import Document
 from fastkml.containers import Folder
+from fastkml.network_link_control import NetworkLinkControl
```
Comment (cleder) [inline fastkml/__init__.py:34]: ordering, run `isort` to fix it

File: fastkml/__init__.py [hunk 11]
```diff
@@ -84,6 +85,7 @@
 
 __all__ = [
     "KML",
+    "NetworkLinkControl",
```
Comment (cleder) [inline fastkml/__init__.py:88]: Ordering, all entries should be ordered alphabetically

File: fastkml/kml.py [hunk 12]
```diff
@@ -59,7 +60,14 @@
...
 logger = logging.getLogger(__name__)
 
-kml_children = Union[Folder, Document, Placemark, GroundOverlay, PhotoOverlay]
+kml_children = Union[
+                Folder,
+                Document,
+                Placemark,
+                GroundOverlay,
+                PhotoOverlay,
+                NetworkLinkControl
```
Comment (cleder) [inline fastkml/kml.py:69]: missing trailing comma.

**Comment by cleder** [comment]
Failing tests in static tests:
```
flake8 fastkml examples docs
fastkml/kml.py:64:17: E126 continuation line over-indented for hanging indent
fastkml/kml.py:70:13: E126 continuation line over-indented for hanging indent
```

**Review by cleder — CHANGES_REQUESTED** [review]

File: fastkml/network_link_control.py [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+
+registry.register(
+    NetworkLinkControl,
+    RegistryItem(
+        ns_ids=("kml",),
+        attr_name="min_refresh_period",
+        node_name="minRefreshPeriod",
+        classes=(float,),
+        get_kwarg=subelement_float_kwarg,
+        set_element=float_subelement,
```
Comment (cleder) [inline fastkml/network_link_control.py:124]: `default=0,` missing

File: fastkml/network_link_control.py [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+registry.register(
+    NetworkLinkControl,
+    RegistryItem(
+        ns_ids=("kml",),
+        attr_name="max_session_length",
+        node_name="maxSessionLength",
+        classes=(float,),
+        get_kwarg=subelement_float_kwarg,
+        set_element=float_subelement,
+    ),
```
Comment (cleder) [inline fastkml/network_link_control.py:136]: `default=-1`

**Pull request merged**